clang/lib/Sema/SemaBounds.cpp

```diff
@@ -2323,29 +2319,46 @@ namespace {
     BoundsExpr *CheckCastExpr(CastExpr *E, CheckedScopeSpecifier CSS,
                               std::pair<ComparisonSet, ComparisonSet>& Facts,
                               SideEffects SE) {
-      // Suppress diagnostics if side effects are disabled.
-      Sema::ExprSubstitutionScope Scope(S, SE == SideEffects::Disabled);
-
-      if (SE == SideEffects::Enabled)
-        CheckDisallowedFunctionPtrCasts(E);
-
       // If the rvalue bounds for e cannot be determined,
       // e may be an lvalue (or may have unknown rvalue bounds).
       BoundsExpr *ResultBounds = CreateBoundsUnknown();
 
-      // Only compute the rvalue bounds for the subexpression if needed.
-      // In future refactoring stages, once TraverseStmt only computes
-      // rvalue bounds once per expression, CheckCastExpr should always
-      // infer and check the bounds for its subexpression.
-      // Before TraverseStmt is refactored, deferring the computation of the
-      // subexpression bounds prevents unacceptable asymptotic complexity.
-      BoundsExpr *SubExprBounds = nullptr;
+      Expr *SubExpr = E->getSubExpr();
+      CastKind CK = E->getCastKind();
 
       bool IncludeNullTerm =
           E->getType()->getPointeeOrArrayElementType()->isNtCheckedArrayType();
+      bool PreviousIncludeNullTerminator = IncludeNullTerminator;
+      IncludeNullTerminator = IncludeNullTerm;
 
-      CastKind CK = E->getCastKind();
-      Expr *SubExpr = E->getSubExpr();
+      // If the lvalue target bounds and lvalue bounds for the
+      // subexpression are needed, they must be computed before
+      // performing potential side effects on the subexpression.
+      bool IsResultCastBounds = (E->getStmtClass() == Stmt::ImplicitCastExprClass ||

```

Comment:
Apparently I didn't save my comment when writing my pull request feedback, and it got lost..  I believe LValueToRvalue casts and ArrayToPointerDecay casts are always implicit casts.  That means that you can structure the code to do an if-test and then move the two if-statements below within it.
```
BoundsExpr *SubExprTarget = nullptr;
BoundsExpr *SubExprLValueBounds = nullptr;
if (E->getStmtClass() == Stmt::ImplicitCastExprClass &&
   !E->getType()->isCheckedPointerPtrType()) {
  if (CK == CK_LValueToRvalue) {
       …
  }
  if (CK == CK_ArrayToPointerDecay) {
   …
  }
}
```